test-data/unit/check-varargs.test

```diff
@@ -1066,3 +1117,166 @@ class C:
+[builtins fixtures/dict.pyi]
+
+[case testUnpackKwargsAssignmentWhenOnlySourceContainsKwargs]
+from typing import NotRequired
+from typing_extensions import Unpack, TypedDict
+
+class Animal(TypedDict):
+    name: str
+
+class Dog(Animal):
+    breed: str
+
+class Example(TypedDict):
+    animal: Animal
+    string: str
+    number: NotRequired[int]
+
+def src(**kwargs: Unpack[Example]): ...
+def dest(*, animal: Dog, string: str, number: int = ...): ...
+
+dest = src  # OK
+
+def dest_with_positional_args(animal: Dog, string: str, number: int = ...): ...
+
+dest_with_positional_args = src  # E: Incompatible types in assignment (expression has type "Callable[[KwArg(Example)], Any]", variable has type "Callable[[Dog, str, int], Any]")
+[builtins fixtures/dict.pyi]
+[typing fixtures/typing-typeddict.pyi]
+
+[case testUnpackKwargsAssignmentWhenOnlyDestinationContainsKwargs]
+from typing_extensions import Unpack, TypedDict
+
+class Animal(TypedDict):
+    name: str
+
+class Dog(Animal):
+    breed: str
+
+def dest(**kwargs: Unpack[Animal]): ...
+def src(name: str): ...
+
+dog: Dog = {"name": "Daisy", "breed": "Labrador"}
+animal: Animal = dog
+
+dest = src  # E: Incompatible function signatures - variable contains **kwargs and the expression does not. See PEP 692 for more details.
+
+[builtins fixtures/dict.pyi]
+[typing fixtures/typing-typeddict.pyi]
+
+[case testUnpackKwargsAssignmentWhenBothSidesContainKwargsTypedWithUnpack]
+from typing_extensions import Unpack, TypedDict
+
+class Animal(TypedDict):
+    name: str
+
+class Dog(Animal):
+    breed: str
+
+def accept_animal(**kwargs: Unpack[Animal]): ...
+def accept_dog(**kwargs: Unpack[Dog]): ...
+
+accept_dog = accept_animal  # OK
+
+def also_accept_animal(**kwargs: Unpack[Animal]): ...
+def also_accept_dog(**kwargs: Unpack[Dog]): ...
+
+also_accept_animal = also_accept_dog  # E: Incompatible types in assignment (expression has type "Callable[[KwArg(Dog)], Any]", variable has type "Callable[[KwArg(Animal)], Any]")
+
+[builtins fixtures/dict.pyi]
+[typing fixtures/typing-typeddict.pyi]
+
+[case testPassingKwargsFromFunctionToFunction]
+from typing_extensions import Unpack, TypedDict
+
+class Animal(TypedDict):
+    name: str
+
+class Dog(Animal):
+    breed: str
+
+def takes_name(name: str):
+    ...
+
+dog: Dog = {"name": "Daisy", "breed": "Labrador"}
+animal: Animal = dog
+
+def foo(**kwargs: Unpack[Animal]):
+    return kwargs["name"]
+
+def bar(**kwargs: Unpack[Animal]):
+    takes_name(**kwargs)
+
+def baz(animal: Animal):
+    takes_name(**animal)
+
+def spam(**kwargs: Unpack[Animal]):
+    baz(kwargs)
+
+foo(**animal)  # OK! foo only expects and uses keywords of 'Animal'.
+
+bar(**animal)  # E: This will fail at runtime because 'breed' keyword will be passed to 'takes_name' as well.
+
+spam(**animal)  # E: Again, 'breed' keyword will be eventually passed to 'takes_name'.
```

Comment:
There core issue here has nothing to do with `Unpack` syntax. Mypy allows this code:
```python
from typing import TypedDict

class Point(TypedDict):
    x: int
    y: int

def foo(*, x: int, y: int) -> None: ...

p: Point
foo(**p)
```
which is unsafe because `p` can be a subtype at runtime. But FWIW I don't think anyone complained about this false negative in ~5 years this feature is available. On the contrary, many people complained about similar checks for other cases where `TypedDict` subtypes at runtime would cause problems, this is why I recently moved them behind new `--extra-checks` flag.